src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -0,0 +1,436 @@
+
+    return dagRequestBuilder
+            .setFlags(flags)
+            .setTimeZoneOffset(timeZoneOffset)
+            .setStartTs(startTs)
+            .build();
+  }
+
+  // See TiDB source code: executor/builder.go:890
+  private DAGRequest buildTableScan() {
+    checkArgument(startTs != 0, "timestamp is 0");
+    DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
+    Executor.Builder executorBuilder = Executor.newBuilder();
+    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+
+
+    // Step1. Add columns to first executor
+    tableInfo.getColumns().forEach(tiColumnInfo -> tblScanBuilder.addColumns(tiColumnInfo.toProto(tableInfo)));
+    executorBuilder.setTp(ExecType.TypeTableScan);
+    tblScanBuilder.setTableId(tableInfo.getId());
+    dagRequestBuilder.addExecutors(executorBuilder.setTblScan(tblScanBuilder));
+    executorBuilder.clear();
+
+    // Step2. Add others
+    // DO NOT EDIT EXPRESSION CONSTRUCTION ORDER
+    // Or make sure the construction order is below:
+    // TableScan/IndexScan > Selection > Aggregation > TopN/Limit
+    TiExpr whereExpr = mergeCNFExpressions(getWhere());
+    if (whereExpr != null) {
+      executorBuilder.setTp(ExecType.TypeSelection);
+      dagRequestBuilder.addExecutors(
+              executorBuilder.setSelection(
+                      Selection.newBuilder().addConditions(whereExpr.toProto())
+              )
+      );
+    }
+
+    if (!getGroupByItems().isEmpty() || !getAggregates().isEmpty()) {
+      Aggregation.Builder aggregationBuilder = Aggregation.newBuilder();
+      getGroupByItems().forEach(tiByItem -> aggregationBuilder.addGroupBy(tiByItem.getExpr().toProto()));
+      getAggregates().forEach(tiExpr -> aggregationBuilder.addAggFunc(tiExpr.toProto()));
+      executorBuilder.setTp(ExecType.TypeAggregation);
+      dagRequestBuilder.addExecutors(
+              executorBuilder.setAggregation(aggregationBuilder)
+      );
+      executorBuilder.clear();
+    }
+
+    if (!getOrderByItems().isEmpty()) {
+      TopN.Builder topNBuilder = TopN.newBuilder();
+      getOrderByItems().forEach(tiByItem -> topNBuilder.addOrderBy(tiByItem.toProto()));
+      executorBuilder.setTp(ExecType.TypeTopN);
+      dagRequestBuilder.addExecutors(executorBuilder.setTopN(topNBuilder));
+      executorBuilder.clear();
+    }
+
+    if (0 != getLimit()) {
+      Limit.Builder limitBuilder = Limit.newBuilder();
+      limitBuilder.setLimit(getLimit());
+      executorBuilder.setTp(ExecType.TypeLimit);
+      dagRequestBuilder.addExecutors(executorBuilder.setLimit(limitBuilder));
+      executorBuilder.clear();
+    }
+
+    getFields().forEach(tiColumnInfo -> dagRequestBuilder.addOutputOffsets(tiColumnInfo.getColumnInfo().getOffset()));
+
+    return dagRequestBuilder
+            .setTimeZoneOffset(timeZoneOffset)
+            .setFlags(flags)
+            .setStartTs(startTs)
+            .build();
+  }
+
+  public TiDAGRequest setTableInfo(TiTableInfo tableInfo) {
```

Comment:
It does not set anything.

Reply:
It sets the private field `tableInfo` to the parameter we passed in. And this method is called in `org.apache.spark.sql.TiStrategy#toCoprocessorRDD`. After that, we can reference `tableInfo` in `com.pingcap.tikv.meta.TiDAGRequest#buildTableScan` to help the building process of DAG `executors`. So it might set something useful, or I misunderstood your comment?

Reply:
my bad. You actually set table info. `requireNonNull` return it.